### PR TITLE
Add option to correct eigenvalues of Shampoo's preconditioner

### PR DESCRIPTION
**Pull request by runame**
This PR adds new functionality via the `use_eigenvalue_correction` argument to `DistributedShampoo`: it enables (approximately) correcting the eigenvalues/running Adam in the eigenbasis of Shampoo's preconditioner. A variation of this method has first been proposed for K-FAC by [1], and [2] note in Appendix B that it can be applied to Shampoo as well; however, they did not present any empirical results or additional discussion. Concurrently to my collaborators and mine work on this method, [3] have empirically evaluated the method on language models, showing promising results.

Notably, when setting `use_eigenvalue_correction=True`, there is typically no need to use learning rate grafting from Adam (`grafting_config=None`) and, when they are available, Adam's optimal `lr`, `betas`, and `weight_decay` should be a good starting point for further tuning. However, the case of `beta2=1.0`, i.e. an AdaGrad-like accumulation, has not been explored yet. Also, in settings where Shampoo would usually graft its learning rate from SGD, grafting might still be beneficial.

This PR might require some refactors and further testing before it can be merged. In the future, more methods for (approximately) computing the eigenvectors of the preconditioner's factor matrices should be supported.

1. [Fast Approximate Natural Gradient Descent in a Kronecker-factored Eigenbasis](https://arxiv.org/abs/1806.03884). Thomas George, César Laurent, Xavier Bouthillier, Nicolas Ballas, Pascal Vincent. NeurIPS, 2018.
2. [Scalable Second-Order Optimization for Deep Learning](https://arxiv.org/pdf/2002.09018.pdf). Rohan Anil, Vineet Gupta, Tomer Koren, Kevin Regan, and Yoram Singer. Tech Report, 2021.
3. [SOAP: Improving and Stabilizing Shampoo using Adam](https://arxiv.org/abs/2409.11321). Nikhil Vyas, Depen Morwani, Rosie Zhao, Itai Shapira, David Brandfonbrener, Lucas Janson, Sham Kakade. Tech Report, 2024.

### CODE REVIEW

File: distributed_shampoo/distributed_shampoo.py [hunk 1]
```diff
@@ -229,16 +240,20 @@ class DistributedShampoo(torch.optim.Optimizer):
...
-        precondition_frequency (int): Frequency for computing root inverse preconditioner. (Default: 1)
+        precondition_frequency (int): Frequency for computing root inverse preconditioner. If use_eigenvalue_correction=True,
+            this is the frequency of the preconditioner eigenbasis computation. (Default: 1)
         start_preconditioning_step (int): Iteration to start computing inverse preconditioner. If -1, uses
             the same value as precondition_frequency. (Default: -1)
         inv_root_override (int, Sequence[int]): Inverse root to use in Shampoo. If a list [l1, l2, ..., lp], then we will
             use -1 / l1 for 1-D tensor (vectors), -1 / l2 for 2-D tensors (matrices), and so on. If the order of the
             tensor exceeds the order of the tensor, reverts to the default value. If 0 is used, uses the default inverse
-            root -1 / (2 * o), where o is the order of the tensor. (Default: 0)
+            root -1 / (2 * o), where o is the order of the tensor. If use_eigenvalue_correction=True, the default is -1 / 2.
```
[the remark this answers is not in the document]
Comment (runame) [reply distributed_shampoo/distributed_shampoo.py:250]: Maybe I misunderstand but when `use_eigenvalue_correction == True` the default `inv_root_override == 0` will lead to the correct exponent -0.5. If `inv_root_override` is set to anything else it works the same as with `use_eigenvalue_correction == False`. What configuration are you concerned about?

File: distributed_shampoo/distributed_shampoo.py [hunk 2]
```diff
@@ -229,16 +240,20 @@ class DistributedShampoo(torch.optim.Optimizer):
...
+            (Default: 0)
         exponent_multiplier (float): Number to be multiplied to the numerator of the inverse root, i.e., eta where the
             exponent is -eta / (2 * p). (Default: 1.0)
-        use_nesterov (bool): Flag for using Nesterov momentum. (default: False)
+        use_eigenvalue_correction (bool): Flag for correcting the eigenvalues/running Adam in the eigenbasis of Shampoo's
```
[the remark this answers is not in the document]
Comment (runame) [reply distributed_shampoo/distributed_shampoo.py:254]: I think this functionality should be fully separate from the grafting configuration because it is in principle completely independent and algorithmically unrelated to grafting. However, we could log a warning in cases where we haven't really explored the eigenvalue correction yet, e.g. `use_eigenvalue_correction == True` and `betas[1] == 1.0` (AdaGrad-style accumulation).

File: distributed_shampoo/distributed_shampoo.py [hunk 3]
```diff
@@ -394,7 +410,8 @@ def __init__(
         if preconditioner_dtype is not None:
             if precision_config is None:
                 precision_config = PrecisionConfig(
-                    factor_matrix_dtype=preconditioner_dtype
+                    factor_matrix_dtype=preconditioner_dtype,
+                    computation_dtype=preconditioner_dtype,
```
[the remark this answers is not in the document]
Comment (runame) [reply distributed_shampoo/distributed_shampoo.py:414]: I only added this because I removed [this code](https://github.com/facebookresearch/optimizers/pull/27/files#diff-1e1a1ed5cd1ee1bf41faa070a43405e5802c862ad9285abb4b6a80665b8d9035L520-L524), but since `preconditioner_dtype` is depreciated, this code will not be needed anyway.

File: distributed_shampoo/utils/shampoo_preconditioner_list.py [hunk 4]
```diff
@@ -524,7 +581,9 @@ def __init__(
 
     @staticmethod
     def _get_inverse_roots_from_override(
-        inv_root_override: Union[int, Sequence[int]], order_list: Tuple[int, ...]
+        inv_root_override: Union[int, Sequence[int]],
+        order_list: Tuple[int, ...],
+        use_eigenvalue_correction: bool,
```
[the remark this answers is not in the document]
Comment (runame) [reply distributed_shampoo/utils/shampoo_preconditioner_list.py:586]: Where would you set the default value for `use_eigenvalue_correction` in this case? For example when `inv_root_override` is a sequence, but a tensor has more dimensions than there elements in the `inv_root_override` sequence, which is currently handled [here](https://github.com/runame/optimizers/blob/eigenvalue-correction/distributed_shampoo/utils/shampoo_preconditioner_list.py#L610-L611)?

File: distributed_shampoo/utils/shampoo_preconditioner_list.py [hunk 5]
```diff
@@ -315,17 +316,22 @@ class ShampooKroneckerFactorsState(OptimizerModule):
     """Shampoo Kronecker Factors (wrapped) for storing in the optimizer state."""
 
     factor_matrices: Tuple[QuantizedTensor, ...]
-    inv_factor_matrices: Tuple[QuantizedTensor, ...]
     factor_matrix_indices: Tuple[str, ...]
     is_factor_matrices_diagonal: Tuple[Tensor, ...] = field(init=False)
+    inv_factor_matrices: Optional[Tuple[QuantizedTensor, ...]] = None
+    factor_matrices_eigenvectors: Optional[Tuple[QuantizedTensor, ...]] = None
```
[the remark this answers is not in the document]
Comment (runame) [reply distributed_shampoo/utils/shampoo_preconditioner_list.py:322]: This makes sense in terms of allowing us to share the code path below. I kept them separate because I couldn't think of a semantically meaningful name for both of them, so I thought it would make the code more readable to have two distinct names.

One general question on simplifying the code path: I have also considered just adding a new class, say `EigenvalueCorrectedShampooPreconditionerList`, that inherits from `ShampooPreconditionerList`. In this case, we could avoid the if-else statements in the methods of `ShampooPreconditionerList` that check if `use_eigenvalue_correction == True`. I haven't thought this fully through, but do you think this would be cleaner?

**Comment by runame** [comment]
@tsunghsienlee Thanks a lot for your comments! I just replied to them.

File: distributed_shampoo/distributed_shampoo.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (runame) [reply distributed_shampoo/distributed_shampoo.py:254]: I still think that the eigenvalue correction should be kept independent from the grafting configuration. I can think more about an alternative interface to the current additional boolean flag that can be passed to the constructor.

File: distributed_shampoo/distributed_shampoo.py [hunk 6]
```diff
@@ -903,6 +943,18 @@ def _update_preconditioners(
...
             )
 
+    @torch.no_grad()
+    def _update_eigenvalue_corrections(
+        self,
+        state_lists: Dict[str, Any],
+        use_eigenvalue_correction: bool,
+    ):
+        # Update eigenvalue correction for the Shampoo preconditioner.
+        if use_eigenvalue_correction:
```
[the remark this answers is not in the document]
Comment (runame) [reply distributed_shampoo/distributed_shampoo.py:953]: Yes, however, the computation of the eigenvalue correction depends on the eigenvector computation, so if their order is switched, the eigenvalue correction will be a step behind. See [my comment]() on the refactor that I did, with this both `PreconditionerList`s will only be called with `update_preconditioners` inside of Shampoo's `step()` function.

File: distributed_shampoo/distributed_shampoo.py [hunk 7]
```diff
@@ -825,6 +850,21 @@ def _compute_root_inverse(
...
+        use_eigenvalue_correction: bool,
+    ) -> None:
+        if compute_root_inverse_or_eigenvectors:
+            if use_eigenvalue_correction:
+                state_lists[SHAMPOO_PRECONDITIONER_LIST].compute_preconditioner_eigenvectors()
+            else:
+                state_lists[SHAMPOO_PRECONDITIONER_LIST].compute_root_inverse()
+                if self._track_root_inv_residuals:
+                    self._compute_and_log_root_inverse_residuals()
+
```
[the remark this answers is not in the document]
Comment (runame) [reply distributed_shampoo/distributed_shampoo.py:867]: I have now merged the root inverse/eigenvector computation into `update_preconditioners`. In this way, Shampoo's `step()` function only has to call this method. Therefore, there are no extra methods that have to be called for different `PreconditionerList`s. In the future, new preconditioners can be implemented by creating a new `PreconditionerList` with its own custom `update_preconditioners` method and Shampoo's step function won't have to be changed.
